### ISSUE TITLE
Error in rendering XRs are not propagated to Claims

### ISSUE THREAD

**Issue by ccremer** [crossplane/crossplane#3262]
Note: The issue of map transform failing is just one example of a case where rendering can fail. The real issue is that any rendering error isn't visible for the user interacting with the claim.

### What happened?

When patching with a Map transform, if the key doesn't exist it silently fails to apply the patch (snippet below).

The composition will stay not ready, with a spec only with 1 condition:
```yaml
  conditions:
    - lastTransitionTime: '2022-08-23T15:09:44Z'
      reason: Creating
      status: 'False'
      type: Ready
```
Only in the logs we can see that it fails applying with an error like this:
```bash
1.6612685509865906e+09	DEBUG	crossplane.events	Warning	{"object": {... "reason": "ComposeResources", "message": "cannot render composed resource from resource template at index 2: cannot apply the patch at index 6: transform at index 0 returned error: map transform could not resolve: key asdf is not found in map"}

tl;dr: "map transform could not resolve: key asdf is not found in map"
```
(this error also is only visible when enabling the debug flag in the args for the crossplane container, e.g. `core start --debug`)

I would expect that the composition receives an additional status condition with the error message, so that it's easier to notice an error when rendering. I consider seeing no errors or status conditions like this by default as a "silent ignore" and therefore a bug, as it's not user friendly to troubleshoot.

(`kubectl describe <the-composition>` does list the warning as an event though, but it's not propagated to the claim.)

### How can we reproduce it?

Composition with patch and a map transform with specific set of keys, like this:
```yaml
- type: FromCompositeFieldPath
  fromFieldPath: spec.parameters.region
  toFieldPath: spec.forProvider.region
  transforms:
    # Any other value from spec.parameters.region result in an silent error.
    - type: map
      map:
        rma: Rümlang
        lpg: Lupfig
```
Then, in the claim's parameter, set a value that doesn't exist in the above map (neither `rma` nor `lpg`).

### What environment did it happen in?
Crossplane version: v1.9

**Comment by ytsarev** [comment]
@negz I think I've seen an associated discussion somewhere(can't find it anymore) where we agreed that we should propagate this to a standard non-debug event stream

**Comment by negz** [comment]
I would expect issues like this to show up as:

* A status condition on the XR
* An event on the XR
* A debug log

What I would not expect is:

* This to be an info log - typically we want folks to look at their infra (e.g. `kubectl describe foo bar`), not Crossplane pod logs, to debug that infra.
* Specific errors like this to be propagated from the XR to the claim. This one is a bit painful for users, but it's idiomatic in Kubernetes when there is an A -> B -> C relationship for A to simply say "B is unhealthy" when B is broken (e.g. think a Deployment with busted ReplicaSets or Pods) rather than showing you what errors B is experiencing.

**Comment by negz** [comment]
I'm slightly second guessing the good first issue label, but I'll leave it there for someone who wants a possibly kinda tricky good first issue. If this is indeed not showing up as a status condition on the XR then it's not working as intended.

**Comment by ccremer** [comment]
Sorry for not using the correct terms, I'm still somewhat learning the terminology.

> * This to be an info log - typically we want folks to look at their infra (e.g. `kubectl describe foo bar`), not Crossplane pod logs, to debug that infra.

I agree

> Specific errors like this to be propagated from the XR to the claim. This one is a bit painful for users, but it's idiomatic in Kubernetes when there is an A -> B -> C relationship for A to simply say "B is unhealthy" when B is broken (e.g. think a Deployment with busted ReplicaSets or Pods) rather than showing you what errors B is experiencing.

I agree this holds true in cases where the user has full access to such resources. However, in a multi-tenant setup our users don't have access to the cluster-scoped resources, they only have claims. (to explain, we can't allow this to users since they would be potentially able to see other resources, and that could lead to unwanted information disclosure without really fine-grained RBAC rules).
Whenever the users create claims, and assuming those won't become ready without any events or conditions, they can't troubleshoot anything if there's something wrong with the spec. If the users (customers in our case) have to involve cluster-admins to troubleshoot issues with incorrect spec then self-service is hard to achieve.

**Comment by jbw976** [comment]
Leaving a breadcrumb to a general community member complaint about silent errors and the challenges of debugging compositions: #2896

**Comment by ccremer** [comment]
I have another case of practically silencing errors:
We have a provider that is using the webhooks feature of Crossplane and it does some validation of the XR's spec.
In my use case I have a claim where a certain property cannot be changed after initial creation. When I do change the property regardless, the validation webhook denies the request to update the managed resource, and the XR is in a state where it wants to apply the new property, but stays "ready".

I observed the following:
- The claim remains in "Ready" condition, the user has no indication (no event in the claim's namespace or status condition) that points out that the property cannot be changed.
- There is an event in the `default` namespace belonging to the XR that contains the webhook's error message.
- The XR still stays in "Ready" condition
- The affected managed resource stays in "Ready" and "Sync" conditions

Without looking at the events in the default namespace (which our users don't have access to) it's basically invisible that Crossplane constantly attempts to update the managed resource, while the claims stays Ready. This can lead the user to believe that the change was successful yet is wondering why the change somehow isn't applied. The user could start wasting time into investigating or open a support request.

I think Crossplane's error handling has room for improvement :)

**Comment by bobh66** [comment]
The error leg records an Event for this condition - https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/reconciler.go#L496 - but there are times when the events don't show up in the describe output so the composite is not reconciling but there is no indication of why outside of the debug logs.

If the intent is to provide the user with enough information to debug the problem without consulting the crossplane logs, maybe there should be a status field on the composite that contains the current error that is blocking reconciliation so that anyone with access to the composite can see the status whether or not the events are available.  Events seem to be unreliable for debugging data that may need to persist over time,

**Comment by jbw976** [comment]
/fresh we'll be focusing on frustrating developer experience issues soon, and this could be one of them :)

**Comment by ytsarev** [comment]
/fresh still the issue from DX perspective

**Comment by jbw976** [comment]
Looks related to https://github.com/crossplane/crossplane/issues/5402 and potentially helped a lot by that design

**Comment by jeanduplessis** [comment]
With [native patch-and-transform deprecated](https://github.com/crossplane/crossplane/pull/5885), and the ability to [communicate errors/events/conditions added to functions](https://github.com/crossplane/crossplane/pull/5450), I will be closing this issue. We recommend using Functions with compositions going forward.

**Comment by pecigonzalo** [comment]
Hey @jeanduplessis (happy to see you around!) - Sorry to revive this, but I believe the linked solution might miss the case of errors in the XRs themselves. The linked resource allow us to propagate and communicate conditions from MRs, but not from the Composite Resource itself I think.

eg.
This is my composite resource status:
```
apiVersion: infra.test.co/v1alpha1
kind: XLambda
metadata:
  name: gonza-test-p5rp6
[...]
status:
  conditions:
  - lastTransitionTime: "2025-03-13T17:06:49Z"
    message: 'cannot compose resources: cannot apply composed resource "LambdaFunction":
      failed to create typed patch object (/gonza-test-p5rp6-k9fd7; lambda.aws.upbound.io/v1beta2,
      Kind=Function): .spec.forProvider.environment.variables.SOME_VAR:
      expected string, got &value.valueUnstructured{Value:1}'
    reason: ReconcileError
    status: "False"
    type: Synced
  - lastTransitionTime: "2025-03-13T17:06:21Z"
    reason: Available
    status: "True"
    type: Ready
```
but the claim does not reflect this in its status
```
apiVersion: infra.test.co/v1alpha1
kind: Lambda
metadata:
  name: gonza-test
[...]
status:
  conditions:
  - lastTransitionTime: "2025-03-13T17:05:37Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
  - lastTransitionTime: "2025-03-13T17:06:21Z"
    reason: Available
    status: "True"
    type: Ready
```

The errors is correct, I should be passing a string, but when applying the resource or inspecting the resource I would not see it has errored out unless I explicitly inspect the composite resource.

**Comment by jeanduplessis** [comment]
👋 @pecigonzalo (small world) – I will try to take another look here soon.
